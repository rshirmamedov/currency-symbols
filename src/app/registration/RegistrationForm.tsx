"use client";

import { Button } from "@/components/Button";
import { Field } from "@/components/Field";
import Link from "next/link";
import { useForm } from "react-hook-form";
import { validateConfirmPassword, validateEmail, validatePassword } from "./validations";
import { sleep } from "@/utils";
import { TIMEOUT } from "@/config";
import { useCallback } from "react";

type RegistrationFormProps = {
    onSubmit: () => void;
};

export const RegistrationForm = (props: RegistrationFormProps) => {
    const {
        register,
        handleSubmit,
        formState: { errors, isSubmitting },
    } = useForm<RegistrationFormFields>();

    const onSubmit = useCallback(async () => {
        await sleep(TIMEOUT);
        props.onSubmit();
    }, [props]);

    return (
        <div data-testid="registration-form" className="max-w-md w-full p-8 border border-gray-300 rounded-md bg-white">
            <h1 className="text-center text-2xl">Registration</h1>
            <div>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <Field
                        inputId="email"
                        type="text"
                        label="Email"
                        placeholder="Enter email"
                        disabled={isSubmitting}
                        error={errors.email?.message}
                        {...register("email", { validate: validateEmail })}
                    />
                    <Field
                        inputId="password"
                        type="password"
                        label="Password"
                        placeholder="Enter password"
                        disabled={isSubmitting}
                        error={errors.password?.message}
                        {...register("password", { validate: validatePassword })}
                    />
                    <Field
                        inputId="confirm-password"
                        type="password"
                        label="Confirm Password"
                        placeholder="Enter confirm password"
                        disabled={isSubmitting}
                        error={errors.confirmPassword?.message}
                        {...register("confirmPassword", { validate: validateConfirmPassword })}
                    />
                    <Button
                        data-testid="register-button"
                        title={isSubmitting ? "Registering..." : "Register"}
                        disabled={isSubmitting}
                    />
                </form>
            </div>
            <div className="text-center p-4 text-sm text-blue-600 hover:text-blue-500">
                <Link href="/login">Login</Link>
            </div>
        </div>
    );
};
