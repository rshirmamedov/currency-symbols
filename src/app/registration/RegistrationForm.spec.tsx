import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { RegistrationForm } from "./RegistrationForm";

describe("RegistrationForm", () => {
    it("renders required email and passwords errors", async () => {
        render(<RegistrationForm onSubmit={jest.fn()} />);

        await userEvent.click(await screen.findByTestId("register-button"));

        expect(await screen.findByTestId('email-field')).toMatchSnapshot();
        expect(await screen.findByTestId('password-field')).toMatchSnapshot();
        expect(await screen.findByTestId('confirm-password-field')).toMatchSnapshot();
    });

    it("renders invalid email error", async () => {
        const user = userEvent.setup();

        render(<RegistrationForm onSubmit={jest.fn()} />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/^password/i);
        const confirmPasswordInput = screen.getByLabelText(/confirm password/i);

        await user.type(emailInput, 'john');
        await user.type(passwordInput, 'password');
        await user.type(confirmPasswordInput, 'password');

        await userEvent.click(await screen.findByTestId("register-button"));

        expect(await screen.findByTestId('email-field')).toMatchSnapshot();
    });

    it("renders passwords match error", async () => {
        const user = userEvent.setup();

        render(<RegistrationForm onSubmit={jest.fn()} />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/^password/i);
        const confirmPasswordInput = screen.getByLabelText(/confirm password/i);

        await user.type(emailInput, "john@doe.com");
        await user.type(passwordInput, "password");
        await user.type(confirmPasswordInput, "Password");

        await userEvent.click(await screen.findByTestId("register-button"));

        expect(await screen.findByTestId('confirm-password-field')).toMatchSnapshot();
    });

    it("disables register button while submitting", async () => {
        const user = userEvent.setup();

        render(<RegistrationForm onSubmit={jest.fn()} />);

        const emailInput = screen.getByLabelText(/email/i);
        const passwordInput = screen.getByLabelText(/^password/i);
        const confirmPasswordInput = screen.getByLabelText(/confirm password/i);

        await user.type(emailInput, 'john@doe.com');
        await user.type(passwordInput, "test123");
        await user.type(confirmPasswordInput, "test123");

        await userEvent.click(await screen.findByTestId("register-button"));

        expect(await screen.findByTestId('register-button')).toMatchSnapshot();
    });

});