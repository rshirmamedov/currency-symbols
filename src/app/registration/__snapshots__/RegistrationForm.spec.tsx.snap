// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm disables register button while submitting 1`] = `
<button
  class="w-full py-2.5 px-4 text-sm rounded text-white bg-blue-600 hover:bg-blue-700 disabled:bg-blue-400 focus:outline-none mt-6"
  data-testid="register-button"
  disabled=""
>
  Registering...
</button>
`;

exports[`RegistrationForm renders invalid email error 1`] = `
<div
  class="py-2"
  data-testid="email-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="email"
  >
    Email
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="email"
    name="email"
    placeholder="Enter email"
    type="text"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Email is invalid
    </div>
  </div>
</div>
`;

exports[`RegistrationForm renders passwords match error 1`] = `
<div
  class="py-2"
  data-testid="confirm-password-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="confirm-password"
  >
    Confirm Password
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="confirm-password"
    name="confirmPassword"
    placeholder="Enter confirm password"
    type="password"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Passwords do not match
    </div>
  </div>
</div>
`;

exports[`RegistrationForm renders required email and passwords errors 1`] = `
<div
  class="py-2"
  data-testid="email-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="email"
  >
    Email
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="email"
    name="email"
    placeholder="Enter email"
    type="text"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Email is required
    </div>
  </div>
</div>
`;

exports[`RegistrationForm renders required email and passwords errors 2`] = `
<div
  class="py-2"
  data-testid="password-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="password"
  >
    Password
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="password"
    name="password"
    placeholder="Enter password"
    type="password"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Password is required
    </div>
  </div>
</div>
`;

exports[`RegistrationForm renders required email and passwords errors 3`] = `
<div
  class="py-2"
  data-testid="confirm-password-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="confirm-password"
  >
    Confirm Password
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="confirm-password"
    name="confirmPassword"
    placeholder="Enter confirm password"
    type="password"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Confirm password is required
    </div>
  </div>
</div>
`;
